File: 이것이코딩테스트다_나동빈/ch16prob36.py
# 편집 거리
# 책 문제와는 다른
# https://www.acmicpc.net/problem/7620
# 으로 대체

INF = int(1e9)
CPY = 0
ADD = 1
DEL = 2
MOD = 3

from collections import deque

str1 = input()
str2 = input()

len1 = len(str1)
len2 = len(str2)

str1 = ' ' + str1
str2 = ' ' + str2

q = deque()
dp = [ [ [INF] * 4 for _ in range(len1 + 1) ] for __ in range(2) ]
sel = True

dp[sel][0][CPY] = 0
dp[sel][0][ADD] = 0
dp[sel][0][DEL] = 0
for i in range(len1 + 1):
    dp[sel][0][MOD] = i

for i2 in range(1, len2 + 1):
    sel = not sel
    dp[sel][0][ADD] = i2
    for i1 in range(1, len1 + 1):

        if str1[i1] == str2[i2]:
            min_cost = min( dp[not sel][i1 - 1] )
            if min_cost != INF:
                dp[sel][i1][CPY] = min_cost

        min_cost = min(dp[not sel][i1])
        if min_cost != INF:
            dp[sel][i1][ADD] = min_cost + 1

        min_cost = min(dp[sel][i1 - 1])
        if min_cost != INF:
            dp[sel][i1][DEL] = min_cost + 1

        min_cost = min(dp[not sel][i1 - 1])
        if min_cost != INF:
            dp[sel][i1][MOD] = min_cost + 1
        
print('min_cost:', min(dp[sel][len1])

# trace minimum cost script
def find_script_op(script_ops, dp, i):
    dp


script_ops = []

File: 이것이코딩테스트다_나동빈/ch16prob36.cpp
# 편집 거리
# 책 문제와는 다른
# https://www.acmicpc.net/problem/7620
# 으로 대체

#define INF 1000000000
#include <iostream>
#include <string>
#include <vector>
using namespace std;

int main() {
    string str1, str2;
    cin >> str1;
    cin >> str2;

    int len1 = str1.length();
    int len2 = str2.length();

    str1.insert(0, " ");
    str2.insert(0, " ");

    vector< tuple< int, vector< tuple<char, char> > > > dp[2];
    bool sel = true;

    for(int i1 = 1; i1 < len1; i1++) {
        vector< tuple
        dp[sel][i1] = [
    }

for i1 in range(1, len1 + 1):
    dp[sel][i1] = [ i1, [('d', str1[i]) for i in range(1, i1 + 1)] ]

for i2 in range(1, len2 + 1):
    sel = not sel
    dp[sel][0] = [ i2, [('a', str2[i]) for i in range(1, i2 + 1)] ]
    for i1 in range(1, len1 + 1):
        min_cost, min_steps = INF, []
        if str1[i1] == str2[i2]:
            cost, steps = dp[not sel][i1 - 1]
            min_cost, min_steps = cost, steps + [('c', str1[i1])]
        else:
            cost, steps = dp[sel][i1 - 1]
            if cost + 1 < min_cost:
                min_cost = cost + 1
                min_steps = steps + [('d', str1[i1])]

            cost, steps = dp[not sel][i1 - 1]
            if cost + 1 < min_cost:
                min_cost = cost + 1
                min_steps = steps + [('m', str2[i2])]

            cost, steps = dp[not sel][i1]
            if cost + 1 < min_cost:
                min_cost = cost + 1
                min_steps = steps + [('a', str2[i2])]
            
        dp[sel][i1] = [min_cost, min_steps]

answer_steps = dp[not sel][len1][1]
for step in answer_steps:
    command, character = step
    print(command, character)

}
